### PR TITLE
refactor(d3-collection): refine and reduce technical debt

### PR DESCRIPTION
**Pull request by tomwanzek**
BREAKING CHANGE:

Set a minimum TypeScript version of 2.4 to be able to
use `object` data type and defaults for generic parameters.

* refactor(): use `object` instead of `any` where basic types should not be permissible.
* chore(): validate and activate `strictNullChecks`
* doc(): complete JSDoc comments
* test(): update tests

Related to #23611 

@gustavderdrache @Ledragon With this PR I am proposing to start using TS 2.3 as the minimum to gain access to relevant language features. Also, when reviewing, please have a good read over the JSDoc comments for `Nest.map(...)`, `Nest.object(...)` and `Nest.entries(...)`. Let me know your thoughts on the "Note" section therein and the related return type considerations. (IMHO we have run along time with these definitions, without having anyone complain about the two edge cases.)

Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

- [x] Provide a URL to documentation or source code which provides context for the suggested changes: <<https://github.com/d3/d3-collection>>
- [x] Increase the version number in the header if appropriate.
- [x] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

EDIT: Update minimum TypeScript version constraint to 2.4

### CODE REVIEW

**Comment by tomwanzek** [comment]
@andy-ms Do you have a take on the following? When I run `npm run lint d3-collection` locally it starts validating and then freezes without error. On Travis the PR passes, but with the below stacktrace. So I am wondering, if there is something going on with the definition changes after all, or it it's a script problem which curiously only affects **d3-collection** I don't see an obvious definitions bug. Thanks in advance!!!

```
<--- Last few GCs --->
[3478:0x2bd9e80]   140771 ms: Mark-sweep 1414.0 (1470.7) -> 1414.0 (1454.7) MB, 2119.2 / 0.0 ms  (+ 0.0 ms in 0 steps since start of marking, biggest step 0.0 ms, walltime since start of marking 2119 ms) last resort GC in old space requested
[3478:0x2bd9e80]   142796 ms: Mark-sweep 1414.0 (1454.7) -> 1414.0 (1454.7) MB, 2024.6 / 0.0 ms  last resort GC in old space requested
<--- JS stacktrace --->
==== JS stack trace =========================================
Security context: 0x1f0395aa5ee1 <JSObject>
    1: createIndexInfo [/home/travis/build/DefinitelyTyped/DefinitelyTyped/node_modules/dtslint/typescript-installs/2.3/node_modules/typescript/lib/typescript.js:~31340] [pc=0x31ba83ba9215](this=0x16b893784f39 <JSGlobal Object>,type=0x25c7a14cd579 <TypeObject map = 0x1fb3bd7d6d11>,isReadonly=0x340526702421 <false>,declaration=0x37d48c6da741 <NodeObject map = 0x35a9e8766019>)
    2: instantiateIn...
FATAL ERROR: CALL_AND_RETRY_LAST Allocation failed - JavaScript heap out of memory
 1: node::Abort() [/home/travis/.nvm/versions/node/v8.9.4/bin/node]
 2: 0x121a2cc [/home/travis/.nvm/versions/node/v8.9.4/bin/node]
 3: v8::Utils::ReportOOMFailure(char const*, bool) [/home/travis/.nvm/versions/node/v8.9.4/bin/node]
 4: v8::internal::V8::FatalProcessOutOfMemory(char const*, bool) [/home/travis/.nvm/versions/node/v8.9.4/bin/node]
 5: v8::internal::Factory::NewFillerObject(int, bool, v8::internal::AllocationSpace) [/home/travis/.nvm/versions/node/v8.9.4/bin/node]
 6: v8::internal::Runtime_AllocateInTargetSpace(int, v8::internal::Object**, v8::internal::Isolate*) [/home/travis/.nvm/versions/node/v8.9.4/bin/node]
 7: 0x31ba8388463d
The command "npm test" exited with 0.
Done. Your build exited with 0.
```

**Comment by andy-ms** [comment]
Hi @tomwanzek, that's definitely a TS bug and I'll have to investigate.

**Comment by tomwanzek** [comment]
@andy-ms OK. It's just peculiar that I am not encountering it in any of the other definitions I quickly checked. Let me know, if anything leads back to the definitions themselves. Cheers, T.

File: types/d3-collection/index.d.ts [hunk 1]
```diff
@@ -2,11 +2,12 @@
 // Project: https://github.com/d3/d3-collection/
 // Definitions by: Tom Wanzek <https://github.com/tomwanzek>, Alex Ford <https://github.com/gustavderdrache>, Boris Yankov <https://github.com/borisyankov>
 // Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
+// TypeScript Version: 2.3
```
Comment (andy-ms) [inline types/d3-collection/index.d.ts]: Out-of-memory bug fixed by setting this to 2.4.

Comment (tomwanzek) [reply types/d3-collection/index.d.ts]: IMHO wouldn't it be better to fix whatever underlying bug causes the issue? There does not seem to be anything too exotic about the definitions. While I can confirm that bumping the TS version to a minimum of 2.4 circumvents the OoM bug, I'm wondering whether we are aiming at the wrong target.

File: types/d3-collection/index.d.ts [hunk 2]
```diff
@@ -16,106 +17,463 @@ export interface Stringifiable {
...
+/**
+ * Returns an array containing the property values of the specified object (an associative array).
+ * The order of the returned array is undefined.
+ *
+ * @param object An object.
+ */
 export function values<T>(object: { [key: string]: T }): T[];
-// TODO: When upgrading definitions to use TS 2.2+, use "object" data type in next line
-export function values(object: any): any[];
+export function values<T = any>(object: object): T[];
```
Comment (gustavderdrache) [inline types/d3-collection/index.d.ts]: Would it be safer to use 2.1's `keyof` support here?

```ts
declare function values<TObject extends object = any>(obj: TObject): Array<TObject[keyof TObject]>;
```

This'll avoid doing silly things like this:

```ts
const dict = { x: 1, y: 2 };

const v = values<string>(point); // oops
```

File: types/d3-collection/index.d.ts [hunk 3]
```diff
@@ -16,106 +17,463 @@ export interface Stringifiable {
...
+ * @param array Array to convert into a map
+ * @param key An optional key function. The functions is invoked for each element in the array being passed
+ * the element's value , it's zero-based index in the array, and the array itself. The function must return a unique string
+ * to be used as the map entry's key.
+ */
+export function map<T>(array: T[], key?: (value: T, i: number, array: T[]) => string): Map<T>;
+/**
+ * Constructs a new map by copying all enumerable properties from the specified object into this map.
+ */
+export function map<T = any>(object: object): Map<T>;
```
Comment (gustavderdrache) [inline types/d3-collection/index.d.ts]: Same `keyof` suggestion applies here.

**Review by gustavderdrache — CHANGES_REQUESTED** [review]
Some questions on the safety of `values()` and `map()`.

Also: do we want to audit this one for `Array<T>` vs. `ReadonlyArray<T>` usage?

File: types/d3-collection/index.d.ts (hunk 1 above)
Comment (andy-ms) [reply types/d3-collection/index.d.ts]: I don't think we would publish a new `typescript@2.3` release to fix a bug like this.

Comment (tomwanzek) [reply types/d3-collection/index.d.ts]: Alright, I thought the reason for the OoM was buried somewhere in the dtslint workings. I noticed that other Definitions (e.g. d3-geo) have had TS 2.3 set for quite some time without any issue. So, if it's TS 2.3 itself, then yep we'll have to bite the bullet and publish as >=TS 2.4.

File: types/d3-collection/index.d.ts (hunk 3 above)
Comment (tomwanzek) [reply types/d3-collection/index.d.ts]: Good old `keyof`. Yeah we had put that on hold back in the day, because the release was too fresh. I don't disagree with your point that the two code pieces you called out where the `object` type is used, are a little sub-optimal from a type-safety perspective.

I just played around with it a little in the context of d3-collection (`keys(...)`, `values(...)`, `entries(...)` and `map(...)`. A couple of things stand out:

* the functionality in d3-collection hits only enumerable properties (including those on `prototype`), `keyof` on the other hand is not restricted to `enumerable`s. So, e.g. calling `keyof A` on a `type A = Array<number | string>` will be vastly different and so will be `A[keyof A]`. Similarly, if one used instances of a class like:
```ts
class TestClass {
    a: 'test';
    b: 123;
    c: 'type';
    private d: 'hidden';
    test() { }
}
```
`keyof ` ignores the private member `d`, but includes the public `test()` method.

* A change management consideration is that so far in all overloaded signatures the generics referred to the data type of the value of the properties. Using the `TObject`-approach diverts from that cognitive model. If we additionally drop the signatures with `xyz<T>(object: { [key: string]: T })` arguments, we might break folks how explicitly set the generic `T`

More to come...

**Comment by tomwanzek** [comment]
As for `ReadonlyArray` I will have a look..

File: types/d3-collection/index.d.ts (hunk 3 above)
Comment (tomwanzek) [reply types/d3-collection/index.d.ts]: ...to continue:
So, in short, I am wondering, if using `keyof` in the case of d3-collection does more harm than good? If we don't use it, at a minimum we would have to keep the overloads for key/value object such as:
```ts
export function values<T>(object: { [key: string]: T }): T[];
```
_prior_ to the final `obj: object` signature, the compiler picks it up first and prevents the "oops" you mentioned for your `dict` example. The key/value signature also will infer a `values(TestClass)` or `values(document)` return type of `any[]` (less specific, but correct).

What the `key/value` signature won't handle alone, are `array`s or more exotic exercises, such as [this actual but somewhat artificial d3-collection test case](https://github.com/d3/d3-collection/blob/1eff7233bac174045331d275e7e0cb793dfdaa0d/test/values-test.js#L11).

Comment (tomwanzek) [reply types/d3-collection/index.d.ts]: @gustavderdrache Correction on my part, the key/value signatures will only be constraining for the `dict` example, if we leave the `object` signature as:
```ts
function values(obj: object): any[];
```

**Comment by tomwanzek** [comment]
@gustavderdrache I have just pushed an interim cleanup commit to address the question of `Array` v `ReadonlyArray` and bump the  TypeScript Constraint to 2.4.

(1) Note that, in the case of the nest operator internally it uses `Array.sort(...)`, so I left the argument constraint to `Array` as is. The other ones are readily reducible to even `ArrayLike` which accepts `ReadonlyArray` as well.
(2) I don't consider the `keyof` discussion from your earlier review comment as closed, so open to your feedback on my earlier response to your review considerations.

Finally, and maybe most importantly, obviously given the Travis Failure, we obviously have to go with a `Typescript 2.4` constraint for `d3`, if we do it for d3-collection. What I am not sure about, is if this constraint cascades further to other DT definitions which depend on d3 or its component modules.

cc @Ledragon

**Comment by tomwanzek** [comment]
As I suspected, if we change to TS 2.4 it cascades quickly even within DT. I just tested the impact of setting `d3` to >=TS 2.4 and now `d3Kit` compilation fails due to conflicting constraints. I suspect, this would not be the only definition outside our "immediate control" that is only on TS 2.3.

Unfortunately, Travis kicks out at the first error, there might be more from other definitions.

@andy-ms your input on ramifications would be more than valued.

**Comment by andy-ms** [comment]
We recently saw a similar error in lodash (#23710). You might consider whether supporting old compilers is more or less important than supporting `ReadonlyArray`.

**Comment by gustavderdrache** [comment]
More out of curiosity than anything else, are there any estimates for usage of each compiler version? It'd be a lot easier to make this decision if we knew that usage was skewed heavily pre-2.4 or not.

**Comment by andy-ms** [comment]
@bowdenk7 might know about version usage.
I suspect, but don't know, that the OoM error is due to the use of ReadonlyArray, since that's what caused it for lodash. The type existed pre-2.3 but there were bugs that caused infinite loops when trying to assign regular arrays to it.

**Comment by tomwanzek** [comment]
@andy-ms thanks for looping @bowdenk7 in on the version usage. It appears  @gustavderdrache and I had the same idea typing in parallel 🤣 

On the `Readonly`, interestingly we had the OoM with only `object`, generics defaults and TS 2.3 constraint. We never ended up using `Readonly`, I went with `ArrayLike` where appropriate.
We recently used `Readonly` in **d3-scale** without any issues (although the definition currently has no min TS constraint)

**Comment by tomwanzek** [comment]
I actually start wondering, I have the sinking feeling that the OoM bug in 2.3 has to do with the signature overloads. Would explain why 2.4 works, since as I recall 2.4 improved some of the inference  in this area.

I will re-create the changes step by step locally from master and see, when dtslint starts to bark at TS2.3.

**Comment by tomwanzek** [comment]
Looks like I found the culprit, nothing essential to what we tried to accomplish. After enforcing `strictNullChecks`, I updated the end of the tests to read:
```ts
type TestL1NestedArrayRollup = Array<{
    key: string;
    value: number;
}>;

type TestL1NestedArrayRollupConservative = Array<{
    key: string;
    value?: number; // Conservatively pretend, value could be undefined
}>;

let testL2NestedArray: TestL2NestedArray;
let testL1NestedArrayRollup: TestL1NestedArrayRollup;
let testL1NestedArrayRollupConservative: TestL1NestedArrayRollupConservative;

testL2NestedArray = nestL2.entries(raw);

num = testL2NestedArray[0].values[0].values[0].yield; // access chain to leaf property
testL1NestedArrayRollupConservative = nestL1Rollup.entries(raw);
// With use-case specific knowledge that a rollup function was applied,
// the existince of value can be asserted with "!"
num = testL1NestedArrayRollupConservative[0].value!; // get rollup value
testL1NestedArrayRollup = nestL1Rollup.entries(raw) as TestL1NestedArrayRollup; // cast as
num = testL1NestedArrayRollup[0].value; // get rollup value
```

Turns out, the "cast as" in the following lines breaks TS 2.3:
```ts
testL1NestedArrayRollup = nestL1Rollup.entries(raw) as TestL1NestedArrayRollup; // cast as
```

I will finalize a new PR with the changes I have not gotten to yet. If I get it to go through smoothly, we have a fresh start on the actual value add of refined d3-collection definitions and I will kill this PR.

**Comment by tomwanzek** [comment]
As the nasty snag has been weeded out, I am closing this PR in favour of #23724.
